File: Sean的leetcode/Leetcode/Flatten 2D Vector.cpp
class Vector2D {
public:
    Vector2D(vector<vector<int>>& vec2d):j(0) {
        start = vec2d.begin();
        end = vec2d.end();
    }

    int next() {
        if (hasNext()) {
            return (*start)[j++];
        }
        else {
            return -1;
        }
    }

    bool hasNext() {
        while (start != end) {
            if ((*start).size() == j) {
                j = 0;
                start++;
            }
            else {
                break;
            }
        }
        return start != end;
    }
    vector<vector<int>>::iterator start, end;
    int j;
};

/**
 * Your Vector2D object will be instantiated and called as such:
 * Vector2D i(vec2d);
 * while (i.hasNext()) cout << i.next();
 */

//https://leetcode.com/discuss/50292/7-9-lines-added-java-and-c-o-1-space
class Vector2D {
public:
    vector<vector<int>>::iterator vcurr, vend;
    int j;
    Vector2D(vector<vector<int>>& vec2d) : j(0) {
        vcurr = vec2d.begin();
        vend = vec2d.end();
    }

    int next() {
        if (hasNext())
            return (*vcurr)[j++];
        return -1;
    }

    bool hasNext() {
        while (vcurr != vend && j == (*vcurr).size())
        {
            vcurr++;
            j = 0;
        }
        return vcurr != vend;
    }
};

/**
 * Your Vector2D object will be instantiated and called as such:
 * Vector2D i(vec2d);
 * while (i.hasNext()) cout << i.next();
 */

File: Sean的leetcode/python/Flatten 2D Vector.py
###
# Design Array
# Time Complexity: hasNext:  O(1) next O(1)
# Space Complexity: O(1)
###
# Your Vector2D object will be instantiated and called as such:
# i, v = Vector2D(vec2d), []
# while i.hasNext(): v.append(i.next())
                                        
                            
class Vector2D(object):

    def __init__(self, vec2d):
        """
        Initialize your data structure here.
        :type vec2d: List[List[int]]
        """
        self.r = 0
        self.c = 0
        self.l = vec2d

    def next(self):
        """
        :rtype: int
        """
        re = self.l[self.r][self.c]
        self.c += 1
        return re

    def hasNext(self):
        """
        :rtype: bool
        """
        while self.r < len(self.l):
            if self.c < len(self.l[self.r]):
                return True
            self.r += 1
            self.c = 0
        return False

# Your Vector2D object will be instantiated and called as such:
# i, v = Vector2D(vec2d), []
# while i.hasNext(): v.append(i.next())   
class Vector2D(object):

    def __init__(self, vec2d):
        """
        Initialize your data structure here.
        :type vec2d: List[List[int]]
        """
        from collections import deque
        whole = [e for v in vec2d for e in v]
        
        self.it = iter(whole)
        self.len = len(whole)

    def next(self):
        """
        :rtype: int
        """
        self.len -= 1
        
        return next(self.it)

    def hasNext(self):
        """
        :rtype: bool
        """
        return self.len != 0

     